### PR TITLE
fix(select): various select fixes (FE-3552, FE-3385, FE-3615)

### PR DESCRIPTION
**Pull request by grabkowski**
### Proposed behaviour
Select visible value should not be cleared on reopening a modal.
Filterable Select should display text value when a default value is set and options are loaded asynchronously.
Filterable Select value should be cleared when the filter does not mach any option.

### Checklist

- [x] Commits follow our style guide
~~- [ ] Screenshots are included in the PR if useful~~
~~- [ ] All themes are supported if required~~
- [x] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
~~- [ ] Storybook added or updated if required~~
~~- [ ] Typescript `d.ts` file added or updated if required~~
~~- [ ] Carbon implementation and Design System documentation are congruent~~

### Testing instructions
fix for value displayed on reopening select when in a modal:
https://codesandbox.io/s/carbon-quickstart-forked-f6ws0
fix for value not displayed on async option load:
https://codesandbox.io/s/carbon-quickstart-forked-yfd7c
fix for value not cleared on incorrect filter:
http://localhost:9001/?path=/docs/design-system-select-filterable--controlled

### CODE REVIEW

File: src/components/select/filterable-select/filterable-select.component.js [hunk 1]
```diff
@@ -208,16 +220,17 @@ const FilterableSelect = React.forwardRef(
...
+          setMatchingText(newValue);
         }
 
         return newValue;
       });
 
       setHighlightedValue(newValue);
-    }, [value, defaultValue, onChange]);
+      // prevent value update on filter change
+      // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (sjohnson-sage) [inline src/components/select/filterable-select/filterable-select.component.js:232]: Why didn't you just include `setMatchingText` as a dependency here?

File: src/components/select/filterable-select/filterable-select.component.js [hunk 2]
```diff
@@ -240,8 +253,9 @@ const FilterableSelect = React.forwardRef(
 
     useEffect(() => {
       setMatchingText(value || defaultValue);
+      // update text value only when children are changing
       // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (sjohnson-sage) [inline src/components/select/filterable-select/filterable-select.component.js:257]: Same question as above but for `defaultValue` `value` and `setMatchingText`?

**Review by sjohnson-sage — COMMENTED** [review]
All looks good to me, just a couple of questions

**Review by mkrds — DISMISSED** [review]

File: src/components/select/filterable-select/filterable-select.component.js (hunk 1 above)
Comment (grabkowski) [reply src/components/select/filterable-select/filterable-select.component.js:232]: I don't want it to be updated on every filterText change

**Review by sjohnson-sage — APPROVED** [review]

File: src/components/select/filterable-select/filterable-select.component.js (hunk 1 above)
Comment (sjohnson-sage) [reply src/components/select/filterable-select/filterable-select.component.js:232]: I didn't read the comment properly 😄

**Review by mkrds — APPROVED** [review]

**Review by sjohnson-sage — APPROVED** [review]

**Review by Alec-Walker-Sage — APPROVED** [review]

**Review by mkrds — APPROVED** [review]

**Review by sjohnson-sage — APPROVED** [review]

**Pull request merged**